Add sleep manager API

0xc0170 opened the pull request:
This goes to feature-hal-spec, based on the sleep manager work that is here #4882 . Review only the last commit (at the moment SHA 34f9f43). This goes to `feature-hal-spec` branch = work in progress

To move further with sleep, I am sending this early to get feedback.

The goal is to have deepsleep locks for those drivers that depend on high speed clocks (it can be that some drivers for a target might support deep sleep but this is not usual, plus speed is limited often).

@bulislaw @c1728p9 @pan-

c1728p9 commented on drivers/CAN.cpp:
```diff
@@ -115,9 +115,11 @@ int CAN::filter(unsigned int id, unsigned int mask, CANFormat format, int handle
 void CAN::attach(Callback<void()> func, IrqType type) {
```
Attach could also be called to replace a callback of the same type. If this is the case then deep sleep shouldn't be locked a second time.

c1728p9 commented on drivers/SerialBase.cpp:
```diff
@@ -73,9 +76,11 @@ void SerialBase::attach(Callback<void()> func, IrqType type) {
     // Disable interrupts when attaching interrupt handler
     core_util_critical_section_enter();
     if (func) {
+        sleep_manager_lock_deep_sleep();
```
Attach could also be called to replace a callback of the same type. If this is the case then deep sleep shouldn't be locked a second time.

c1728p9 commented on drivers/Ticker.cpp:
```diff
@@ -26,6 +26,7 @@ void Ticker::detach() {
     core_util_critical_section_enter();
     remove();
     _function = 0;
+    sleep_manager_unlock_deep_sleep();
```
Detach could be called while already detached (like during object destruction) which could cause an underflow.

c1728p9 commented on drivers/Timer.cpp:
```diff
@@ -41,6 +42,7 @@ void Timer::stop() {
     core_util_critical_section_enter();
     _time += slicetime();
     _running = 0;
+    sleep_manager_unlock_deep_sleep();
```
Unlock should only be called if the timer was running.

0xc0170 commented:
@c1728p9 Added fixes for the last review.

c1728p9 commented on drivers/I2C.cpp:
```diff
@@ -125,6 +125,7 @@ void I2C::unlock() {
 {
     lock();
+    sleep_manager_lock_deep_sleep();
     if (i2c_active(&_i2c)) {
```
If I2C is already active, you aren't starting a new transfer.  You should probably either unlock before returning or add the lock until after the active check.

c1728p9 commented on drivers/CAN.cpp:
```diff
@@ -115,9 +115,14 @@ int CAN::filter(unsigned int id, unsigned int mask, CANFormat format, int handle
         _irq[(CanIrqType)type] = func;
         can_irq_set(&_can, (CanIrqType)type, 1);
     } else {
+        sleep_manager_unlock_deep_sleep();
```
Same problem could occur with the else case - it is valid to call attach with null multiple times in a row. Maybe as a general strategy, each interrupt type could be locked when `donothing` is replaced, and unlocked when `donothing` is set. A similar change should also be made for SerialBase.

0xc0170 commented:
@c1728p9 Fixed, I'll rebase once all approved to clean the history

0xc0170 commented:
This now should be feature complete (testing ongoing, plus reviews so expect me to rebase). 

- add sleep manager API
- deep sleep locks in drivers (Serial, SPI, etc)
- rtos idle loop locks deep sleep (backward compability) for now

This is still against feature hal spec branch. I'll send tests as separate patch soon.

Quickly tested with rtos tests for K64F and GCC ARM - all OK

0xc0170 commented:
retest uvisor

0xc0170 commented:
/morph test

0xc0170 commented:
/morph test

studavekar commented:
/morph test

0xc0170 commented:
I added DeepSleepLock RAII object 

Please review

0xc0170 commented on drivers/CAN.h:
```diff
@@ -24,6 +24,7 @@
 #include "platform/Callback.h"
 #include "platform/PlatformMutex.h"
 #include "platform/NonCopyable.h"
+#include "platform/mbed_sleep.h"
```
@nvlsianpu thanks for the review, the PR updated , all should be fixed.

0xc0170 commented:
/morph test

0xc0170 commented:
I changed a destination branch to master. Rebased it on top of the latest master.

To make this fully backward compatible, I can remove changes in sleep API (invoking sleep manager there). This would be just an addition of sleep manager that is not yet used anywhere (idle loop invokes sleep that is still just hal_sleep wrapper). Let me know what you think

I added tests to test sleep manager API, run on k64f and nucleo f429 with various toolchains.

0xc0170 commented:
@tommikas Can you help here please , the jenkins CI reports:

```
[wlan-8022] 10:29:21.659 | TC     MainThread: Test Case fails because of: D1 CMD timeout: ifconfig wlan0 --ap systest --pwd RaaSysT35T --security WPA/WPA2 up
```

I can't locate what is wrong. How to reproduce this issue? The behaviour of idle loop is the same, so should not affect the tests.

tommikas commented:
Have you tested the ESP8266 driver with these changes? Based on the logs the interface `connect()` call isn't returning quickly enough if at all.

0xc0170 commented:
> Have you tested the ESP8266 driver with these changes? Based on the logs the interface connect() call isn't returning quickly enough if at all.
 
The code looks fine but yes it fails, thanks to @bulislaw for providing me the tracebacks, I was able to identify the issue. It seems that Callback has a bug that I am now investigating. I'll create a new issue and possibly a fix.

0xc0170 commented:
#5017 reported, waiting for a fix, that should unblock this PR

0xc0170 commented:
Rebased, and the SerialBase bug with callback should be fixed now (the class uses Callback differently - does check if it was assigned or not).

0xc0170 commented:
/morph test

0xc0170 commented:
I fixed the above failure for tests, the NOT SUPPORTED was not placed properly, thus mbed-cli reported a failure. Now it should be skipped, tested locally.

0xc0170 commented:
/morph test

0xc0170 commented:
All CI are green, we should be good to go. I'll write documentation for this new feature.

Note, we need to emphasize this in the release notes, as whoever uses sleep()/deepsleep(), an update might change the behavior (**breaking  change**) they will need to update their libraries.

If all happy, this will be integrated today.

I would like to run nightly on this one, but not required.

0xc0170 commented:
I had to fix CAN attach, thus rebased the last commit, will require rerun CI.

theotherjimmy commented:
@0xc0170 If we need to rerun and we need this in 5.6.0, should we just trigger it now?

/morph test

0xc0170 commented:
> @0xc0170 If we need to rerun and we need this in 5.6.0, should we just trigger it now?

The last 3 run failed, as seems something wrong with CI, waiting for Shrikant

0xc0170 commented:
Rebased (removed HAL docs changes, will come as separate PR), tested with nucloe f429, all good.

studavekar commented:
/morph test

studavekar commented:
Jobs seems have completed with results : http://mbed-ci-master-2.austin.arm.com:8081/job/pr_pipeline/1247/

theotherjimmy commented:
Awesome! LGTM.

c1728p9 commented on drivers/Ticker.h:
```diff
@@ -102,6 +103,10 @@ class Ticker : public TimerEvent, private NonCopyable<Ticker> {
     void attach_us(Callback<void()> func, us_timestamp_t t) {
+        // lock only for the initial callback setup
+        if (!_function) {
+            sleep_manager_lock_deep_sleep();
```
How does the sleep manager get unlocked?

theotherjimmy replied:
Generally, `sleep_manager_unlock_deep_sleep`, but you're asking for the specific unlock for this call to lock right?

c1728p9 commented on drivers/Timer.cpp:
```diff
@@ -40,6 +41,9 @@ void Timer::start() {
     core_util_critical_section_enter();
     _time += slicetime();
+    if (_running) {
+        sleep_manager_unlock_deep_sleep();
```
You might want to unlock this on object destruction.

c1728p9 commented on platform/DeepSleepLock.h:
```diff
@@ -0,0 +1,67 @@
+
+    /** Mark the start of a locked deep sleep section
+     */
+    void lock()
```
Is there any reason to expose lock and unlock?

sg- commented:
Nice change removing the donothing. Looks like there is one straggler left in the codebase. https://github.com/ARMmbed/mbed-os/search?utf8=%E2%9C%93&q=callback%28donothing%29&type=

0xc0170 commented:
> Nice change removing the donothing. Looks like there is one straggler left in the codebase. https://github.com/ARMmbed/mbed-os/search?utf8=%E2%9C%93&q=callback%28donothing%29&type=
 
It was left there as donothing replacement in this PR was required for making new API working (compare if a callback was attached or not). Thanks for checking ! Yes, I'll send PR shortly to align also InterruptIn

jeromecoutant commented:
Hi
In tests-mbed_drivers-lp_timeout and tests-mbed_hal-lp_ticker, we are still using deepsleep function which is deprecated.
Maybe you should update these tests ?
Thx

0xc0170 commented:
> In tests-mbed_drivers-lp_timeout and tests-mbed_hal-lp_ticker, we are still using deepsleep function which is deprecated.
Maybe you should update these tests ?

Yes, I'll do

nvlsianpu commented:
Apparently looks like this path broke nrf52_Dk (nrf52840 as well) spi (tests-api-spi). Commit  cb4e9b32a2933553630ff1ed40f9bdb77e51fda2 introduced such a change. Even if i comment off sleep instruction it still failed.

nvlsianpu commented:
@0xc0170, @pan-- ^^

0xc0170 commented:
@nvlsianpu  I do not fully get even if you comment out sleep, then it still fails? Can you create an issue with details, I would like to reproduce it locally .

nvlsianpu commented:
@0xc0170 https://github.com/ARMmbed/mbed-os/issues/5297
